### PR TITLE
GFORMS-740 - dmsFormId length validations

### PR DESCRIPTION
**Pull request by SandipKhedekar**

### CODE REVIEW

File: test/uk/gov/hmrc/gform/sharedmodel/formtemplate/generators/DestinationGen.scala [hunk 1]
```diff
@@ -28,7 +28,12 @@ trait DestinationGen {
       r <- Gen.alphaNumStr
     } yield DestinationId(f.toString + r)
 
-  def dmsFormIdGen: Gen[String] = PrimitiveGen.nonEmptyAlphaNumStrGen
+  def dmsFormIdGen: Gen[String] = PrimitiveGen.nonEmptyAlphaNumStrGen.map { dmsFormId =>
+    if (dmsFormId.length > 12)
```
Comment (girishkamat) [inline test/uk/gov/hmrc/gform/sharedmodel/formtemplate/generators/DestinationGen.scala:32]: I dont think we need this check. Just replace with `dmsFormId.take(12)`

File: app/uk/gov/hmrc/gform/formtemplate/FormTemplatesControllerRequestHandler.scala [hunk 2]
```diff
@@ -157,7 +162,35 @@ object FormTemplatesControllerRequestHandler {
         case (None, Some(_), None, None)       => JsSuccess(())
       }
 
-    sectionValidations andKeep jsonValue.transform(
+    val dmsFormIdValidations: JsResult[Unit] =
+      (jsonValue \ "destinations").toOption match {
+        case Some(id) =>
```
Comment (girishkamat) [inline app/uk/gov/hmrc/gform/formtemplate/FormTemplatesControllerRequestHandler.scala:167]: the variable id could be renamed to destinationsJsArray or something else

File: app/uk/gov/hmrc/gform/formtemplate/FormTemplatesControllerRequestHandler.scala [hunk 3]
```diff
@@ -157,7 +162,35 @@ object FormTemplatesControllerRequestHandler {
...
+          val errorMessages = id
+            .as[JsArray]
+            .value
+            .map { jv =>
+              ((jv \ "id").as[String], (jv \ "type").as[String], (jv \ "dmsFormId").toOption.map(_.as[String]))
+            }
+            .map {
+              // format: off
+              case (_, "hmrcDms", Some(dmsFormId))
+                  if dmsFormId.toString.length > 0 && dmsFormId.toString.length <= 12 => ""
```
Comment (girishkamat) [inline app/uk/gov/hmrc/gform/formtemplate/FormTemplatesControllerRequestHandler.scala:177]: Probably a good idea to trim the value before checking as a space seems to get through fine `"dmsFormId": " "`

File: app/uk/gov/hmrc/gform/formtemplate/FormTemplatesControllerRequestHandler.scala [hunk 4]
```diff
@@ -157,7 +162,35 @@ object FormTemplatesControllerRequestHandler {
...
 
-    sectionValidations andKeep jsonValue.transform(
+    val dmsFormIdValidations: JsResult[Unit] =
+      (jsonValue \ "destinations").toOption match {
+        case Some(id) =>
+          val errorMessages = id
+            .as[JsArray]
+            .value
+            .map { jv =>
+              ((jv \ "id").as[String], (jv \ "type").as[String], (jv \ "dmsFormId").toOption.map(_.as[String]))
```
Comment (girishkamat) [inline app/uk/gov/hmrc/gform/formtemplate/FormTemplatesControllerRequestHandler.scala:172]: `(jv \ "dmsFormId").toOption.map(_.as[String])` could be simplified to `(jv \ "dmsFormId").asOpt[String]`

File: test/uk/gov/hmrc/gform/formtemplate/FormTemplatesControllerRequestHandlerTest.scala [hunk 5]
```diff
@@ -789,4 +939,186 @@ class FormTemplatesControllerRequestHandlerTest extends WordSpec with Matchers w
        |    "fields": []
        |  }
        |}""".stripMargin
```
Comment (girishkamat) [inline test/uk/gov/hmrc/gform/formtemplate/FormTemplatesControllerRequestHandlerTest.scala:941]: all the valid and invalid json methods below could be simplified, to a single json and just replacing the dmsFormId part with a passed in value, which could be an Option. Reduces a lot of copy paste also makes it simpole to updated the templates change in future

File: test/uk/gov/hmrc/gform/sharedmodel/formtemplate/generators/DestinationGen.scala (hunk 1 above)
Comment (SandipKhedekar) [reply test/uk/gov/hmrc/gform/sharedmodel/formtemplate/generators/DestinationGen.scala]: Yes. Changed.

File: app/uk/gov/hmrc/gform/formtemplate/FormTemplatesControllerRequestHandler.scala (hunk 2 above)
Comment (SandipKhedekar) [reply app/uk/gov/hmrc/gform/formtemplate/FormTemplatesControllerRequestHandler.scala:167]: I believe variable id **dmsFormIdValidations** is more apporiprate. It's inline with othera   **sectionValidations andKeep dmsFormIdValidations andKeep jsonValue.transform(**

File: app/uk/gov/hmrc/gform/formtemplate/FormTemplatesControllerRequestHandler.scala (hunk 3 above)
Comment (SandipKhedekar) [reply app/uk/gov/hmrc/gform/formtemplate/FormTemplatesControllerRequestHandler.scala]: Good catch. Addressed.

File: test/uk/gov/hmrc/gform/formtemplate/FormTemplatesControllerRequestHandlerTest.scala (hunk 5 above)
Comment (SandipKhedekar) [reply test/uk/gov/hmrc/gform/formtemplate/FormTemplatesControllerRequestHandlerTest.scala:883]: Yes. Changed.

File: app/uk/gov/hmrc/gform/formtemplate/FormTemplatesControllerRequestHandler.scala (hunk 4 above)
Comment (SandipKhedekar) [reply app/uk/gov/hmrc/gform/formtemplate/FormTemplatesControllerRequestHandler.scala]: In my code, I have addressed the possibility of missing **dmsFormId** And I am using **missingDmsFormId** message in this case.

File: app/uk/gov/hmrc/gform/formtemplate/FormTemplatesControllerRequestHandler.scala (hunk 2 above)
Comment (girishkamat) [reply app/uk/gov/hmrc/gform/formtemplate/FormTemplatesControllerRequestHandler.scala:167]: case Some(id), here id represents an array of destinations and not just a single id. it cause a bit of confusion when I read it for a the first time.

File: app/uk/gov/hmrc/gform/formtemplate/FormTemplatesControllerRequestHandler.scala (hunk 4 above)
Comment (girishkamat) [reply app/uk/gov/hmrc/gform/formtemplate/FormTemplatesControllerRequestHandler.scala]: You dont need to check for missingDmsFormId, its already taken care when you try (jv \ "dmsFormId").as[String]. Why add new code when existing checks already work?

**Review by girishkamat — APPROVED** [review]

**Pull request merged**